### PR TITLE
Add sequence mask for grpo

### PR DESCRIPTION
**Pull request by garyzhang99**
## Description

Here we implement the sequence masking formula introduced in [DeepSeek V3.2 paper](https://huggingface.co/deepseek-ai/DeepSeek-V3.2/blob/main/assets/paper.pdf)

image

Since the $\delta$ hyperparameter is not given a recommended value in the original paper, we further tested the influence of $\delta$. It seems that setting the range from `0.05` to `0.2` is reasonable.

image

## Checklist

Please check the following items before code is ready to be reviewed.

- [ ]  Code has passed all tests
- [ ]  Docstrings have been added/updated in Google Style
- [ ]  Documentation has been updated
- [ ]  Code is ready for review

### CODE REVIEW

File: trinity/algorithm/policy_loss_fn/ppo_policy_loss.py [hunk 1]
```diff
@@ -36,6 +38,8 @@ def __init__(
         assert self.clip_range_low is not None, "clip_range_low must be specified."
         assert self.clip_range_high is not None, "clip_range_high must be specified."
         self.loss_agg_mode = loss_agg_mode
+        self.enable_sequence_masking = enable_sequence_masking
+        self.delta = delta
```
Comment (yanxi-chen) [inline trinity/algorithm/policy_loss_fn/ppo_policy_loss.py]: The name "delta" feels too generic. Maybe consider "delta_sequence_masking"?

File: trinity/algorithm/policy_loss_fn/ppo_policy_loss.py [hunk 2]
```diff
@@ -51,6 +55,33 @@ def __call__(  # type: ignore
...
+            should_mask = has_negative_advantage & exceeds_kl_threshold
+            sequence_mask = (~should_mask).float()
+
+            metrics_seq_mask = {
+                "seq_mask/masked_tokens": should_mask.float().sum().item()
+                / (action_mask.sum().item() + 1e-10),
+                "seq_mask/mean_sequence_kl": sequence_kl.mean().detach().item(),
+            }
+        else:
+            metrics_seq_mask = {}
```
Comment (yanxi-chen) [inline trinity/algorithm/policy_loss_fn/ppo_policy_loss.py]: Remove these two lines

File: trinity/algorithm/policy_loss_fn/ppo_policy_loss.py [hunk 3]
```diff
@@ -66,13 +97,19 @@ def __call__(  # type: ignore
...
+            pg_losses = pg_losses * sequence_mask
+
         pg_loss = aggregate_loss(pg_losses, action_mask, loss_agg_mode=self.loss_agg_mode)
         metrics = {
             "pg_clipfrac": pg_clip_frac.detach().item(),
             "pg_clipfrac_lower": pg_clipfrac_lower.detach().item(),
             "ppo_kl": ppo_kl.detach().item(),
             "pg_loss": pg_loss.detach().item(),
         }
+        metrics.update(metrics_seq_mask)
```
Comment (yanxi-chen) [inline trinity/algorithm/policy_loss_fn/ppo_policy_loss.py]:
```suggestion
        if self.enable_sequence_masking:
            metrics.update(metrics_seq_mask)
```

File: trinity/algorithm/policy_loss_fn/ppo_policy_loss.py [hunk 4]
```diff
@@ -51,6 +55,33 @@ def __call__(  # type: ignore
         ratio = torch.exp(negative_approx_kl)
         ppo_kl = masked_mean(-negative_approx_kl, action_mask)
 
+        # Compute sequence masking if enabled
+        sequence_mask = torch.ones_like(advantages)
```
Comment (yanxi-chen) [inline trinity/algorithm/policy_loss_fn/ppo_policy_loss.py]: Move this line into the "if" below

File: trinity/algorithm/policy_loss_fn/ppo_policy_loss.py [hunk 5]
```diff
@@ -66,13 +97,19 @@ def __call__(  # type: ignore
             torch.gt(clip_pg_losses1, pg_losses3) * (advantages < 0).float(), action_mask
         )
         pg_losses = torch.where(advantages < 0, clip_pg_losses2, clip_pg_losses1)
+
+        # Apply sequence mask to the losses
+        if self.enable_sequence_masking:
+            pg_losses = pg_losses * sequence_mask
```
Comment (yanxi-chen) [inline trinity/algorithm/policy_loss_fn/ppo_policy_loss.py]: If possible, merging the previous lines related to sequence masking into this "if" will make the code cleaner.
